cmd/webhook/main.go

```diff
@@ -33,15 +35,64 @@ func main() {
 
 	glog.Infof("starting mutating admission controller for network resources injection")
 
+	keyPair, err := webhook.NewTlsKeypairReloader(*cert, *key)
+	if err != nil {
+		glog.Fatalf("error load certificate: %s", err.Error())
+	}
+
 	/* init API client */
 	webhook.SetupInClusterClient()
 
-	/* register handlers */
-	http.HandleFunc("/mutate", webhook.MutateHandler)
+	go func() {
+		/* register handlers */
+		var httpServer *http.Server
+		http.HandleFunc("/mutate", webhook.MutateHandler)
+
+		/* start serving */
+		httpServer = &http.Server{
+			Addr: fmt.Sprintf("%s:%d", *address, *port),
+			TLSConfig: &tls.Config{
+				GetCertificate: keyPair.GetCertificateFunc(),
+			},
+		}
 
-	/* start serving */
-	err := http.ListenAndServeTLS(fmt.Sprintf("%s:%d", *address, *port), *cert, *key, nil)
+		err := httpServer.ListenAndServeTLS("", "")
+		if err != nil {
+			glog.Fatalf("error starting web server: %v", err)
+		}
+	}()
+
+	/* watch the cert file and restart http sever if the file updated. */
+	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
-		glog.Fatalf("error starting web server: %s", err)
+		glog.Fatalf("error starting fsnotify watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	for {
+		watcher.Add(*cert)
+		watcher.Add(*key)
+
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				continue
+			}
+			glog.Infof("watcher event: %v", event)
+			if event.Op & fsnotify.Create == fsnotify.Create ||
+				event.Op & fsnotify.Rename == fsnotify.Rename ||
+				event.Op & fsnotify.Remove == fsnotify.Remove ||
+				event.Op & fsnotify.Write == fsnotify.Write {
```

Comment:
question - why do we watch for Remove and Rename? and not for Chmod (the only Op that's left out here)
it could be a bit simpler if we handled any type of operation (including chmod) or only the ones that we need to

But if we need precisely this set of Ops than that's fine.

Also, nit comment on the readability/simplicity - Op is uint32 so maybe we could simplify it a bit and do something like this:
```
mask := fsnotify.Create | fsnotify.Rename | fsnotify.Remove | fsnotify.Write
if (event.Op & mask) != 0 {
    ...
}
```

Reply:
I tried to capture all the possible events during testing, and found that depending on different operations(rename, move, delete, create etc), it triggers various event, so I'd like to capture all of them and reload the certificate(reloading is cheap, it doesn't require much cycles);
I will also add chmod and use your simplified code:
```
mask := fsnotify.Create | fsnotify.Rename | fsnotify.Remove | fsnotify.Write | fsnotify.Chmod
if (event.Op & mask) != 0 {
    ...
}
```